lib/web/desktop.go

```diff
@@ -111,14 +111,24 @@ func (h *Handler) createDesktopConnection(
 		return err
 	}
 
-	q := r.URL.Query()
-	username := q.Get("username")
-	if username == "" {
-		return sendTDPError(trace.BadParameter("missing username"))
+	username, err := readUsername(r)
+	if err != nil {
+		return sendTDPError(err)
 	}
-
 	log.Debugf("Attempting to connect to desktop using username=%v\n", username)
 
+	// Read the tdp.ClientScreenSpec from the websocket.
+	// This is always the first thing sent by the client.
+	// Certificate issuance may rely on the client sending
+	// a subsequent tdp.MFA message, hence we need to make
+	// sure that this message has been read from the wire
+	// beforehand.
+	screenSpec, err := readClientScreenSpec(ws)
```

Comment:
So we start exchanging data (screen size) before we've done MFA?

Is this new behavior or has this always been the case?

Reply:
This was always the case, previous to the ironrdp transition we were sending the screen spec as a url param (same as username).

Reply:
Sounds like a behavior change to me, since we now allow TDP messages to be exchanged prior to passing MFA.

How is this going to work with a v15 proxy and a v14 w_d_s?

Reply:
The change is purely with how the web client communicates with the proxy.

v14:
1. web client sends username and screen spec as url params to the proxy
2. proxy saves these
3. proxy goes through mfa ceremony
4. proxy sends username and then screen spec as tdp messages to the wds

v15:
1. web client sends username as url param **and screen spec as tdp message** to the proxy
2. proxy saves these
3. proxy goes through mfa ceremony
4. proxy sends username and then screen spec as tdp messages to the wds

so as long as the web client is always communicating with the proxy that served it, this piece of the process will work.

Reply:
Gotcha. Thanks for clarifying.

> So as long as the web client is always communicating with the proxy that served it

We generally shouldn't make this assumption (it's not always true), but it's okay for now.